### PR TITLE
Instadapp: add Liquity resolver (read-only positional data)

### PR DESCRIPTION
**Pull request by edmulraney**
As per Instadapp integration guide: https://www.notion.so/Instadapp-Integration-98d638a7917344afb4e2be63a09cf9a6

We're required to add a read-only wrapper over the protocol.

I've added five public readers:
- Trove position data (collateral, debt, ICR)
- Stability deposit position data (deposit, ethGain, lqtyGain)
- Stake position data (amount, ethGain, lusdGain)
- Overall position data (Trove, StabilityDeposit, Stake)
- Overall system state (borrowFee, ethTvl, TCR, isInRecoveryMode)

To run the tests (`npm test`) you need to add an `ALCHEMY_API_KEY` to your environment variables, and comment out the Uniswap test file since it's broken and kills the test suite otherwise.

### CODE REVIEW

File: contracts/flusher/balances.sol [hunk 1]
```diff
@@ -28,7 +28,7 @@ contract Resolver {
             tokensBal[i] = Balances({
                 flusher: flushers[i],
                 balance: bals,
-                isDeployed: isContractDeployed(flushers[i]);
+                isDeployed: isContractDeployed(flushers[i])
```
Comment (edmulraney) [inline contracts/flusher/balances.sol:31]: Project didn't build because of this syntax error. Not related to Liquity.

File: contracts/protocols/mainnet/liquity.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    StabilityDeposit stability;
+    Stake stake;
+  }
+}
+
+
+contract Resolver is Helpers {
+  function getTrove(address owner) public view returns (Trove memory) {
+    (uint debt, uint collateral, uint _, uint __) = troveManager.getEntireDebtAndColl(owner);
+    uint price = priceFeed.lastGoodPrice();
```
Comment (danielattilasimon) [inline contracts/protocols/mainnet/liquity.sol]: I wonder how important it is to use an up-to-date price here. `lastGoodPrice` can lag behind the actual Chainlink (or Tellor) price by quite a bit at times when the frequency of Liquity transactions is low.

Of course we can't just call `fetchPrice()` from a view function, but maybe we could create a read-only twin of the `PriceFeed` contract that looks at the public state of `PriceFeed` and gets the price according to the same logic, but makes no state changes...

Or if that's too far fetched, we could just use a single underlying oracle (Chainlink) and read it directly.

File: contracts/protocols/mainnet/liquity.sol [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  function getPosition(address owner) external view returns (Position memory) {
+    Trove memory trove = getTrove(owner);
+    StabilityDeposit memory stability = getStabilityDeposit(owner);
+    Stake memory stake = getStake(owner);
+    return Position(trove, stability, stake);
+  }
+
+  function getSystemState() external view returns (System memory) {
+    uint borrowFee = troveManager.getBorrowingRateWithDecay();
+    uint ethTvl = add(activePoolAddress.balance, defaultPoolAddress.balance);
```
Comment (bingen) [inline contracts/protocols/mainnet/liquity.sol]: I would use the getters `getETH` in those contracts. In some extreme edge cases the value may differ.

File: test/liquity.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
+const { expect } = require("chai");
+const hardhatConfig = require("../hardhat.config");
+const { BigNumber } = hre.ethers;
+
+// Deterministic block number to run these tests from on forked mainnet. If you change this, tests will break.
+const BLOCK_NUMBER = 12478959;
+
+// Liquity user with a Trove, Stability deposit, and Stake
+const JUSTIN_SUN_ADDRESS = "0x903d12bf2c57a29f32365917c706ce0e1a84cce3";
```
Comment (bingen) [inline test/liquity.js:9]: :joy:

**Review by bingen — APPROVED** [review]

File: contracts/protocols/mainnet/liquity.sol (hunk 2 above)
Comment (bingen) [reply contracts/protocols/mainnet/liquity.sol]: That’s a good question. Calling Chainlik may become less reliable in some circumstances too. What are the resulting ICR and TCR values used for?

Comment (edmulraney) [reply contracts/protocols/mainnet/liquity.sol]: Great point @danielattilasimon. I guess we need to use the price that our Liquity protocol would present to the user. Creating a read only (view) `getPrice()` would simplify this significantly. 

We want to show same value a user would see inside a Liquity frontend, since this wrapper will be used by Instadapp for their frontend integration of Liquity.

Comment (danielattilasimon) [reply contracts/protocols/mainnet/liquity.sol]: I guess one more way to go would be to get rid of the `view` modifier on `getTrove()`, which would let you call `fetchPrice()` on the `PriceFeed`. However, this may not work for Instadapp (e.g. if there's a hard requirement that all resolver functions must be constant). For example if they have some helper contract that `STATICCALL`s resolver functions, then this won't work as the call will revert (even if `eth_call`ed).

Comment (edmulraney) [reply contracts/protocols/mainnet/liquity.sol]: Added eth oracle price as a param: https://github.com/liquity/dsa-resolvers/pull/1/commits/34b7c390251164a64db77d24589af1620c528c41

File: contracts/protocols/mainnet/liquity.sol (hunk 3 above)
Comment (edmulraney) [reply contracts/protocols/mainnet/liquity.sol]: https://github.com/liquity/dsa-resolvers/pull/1/commits/34b7c390251164a64db77d24589af1620c528c41
